cli/flox/src/commands/init/node.rs

```diff
@@ -257,88 +218,478 @@ impl Node {
+        if !path.exists() {
+            debug!(path = %path.display(), ".nvmrc didn't exist");
+            return Ok(None);
+        }
+
+        let contents = fs::read_to_string(path);
+        let debug_contents = contents
+            .as_ref()
+            .map(|s| s.as_str())
+            .unwrap_or("read_error");
+        debug!(contents = debug_contents, "read .nvmrc");
+
+        contents.context("failed to read .nvmrc file").map(Some)
+    }
+
+    /// Returns whether this state is even valid before we start looking for
+    /// compatible versions or trying to decide which packages to install.
+    fn is_valid_state(&self) -> bool {
+        // We need _some_ kind of information, and at the very least one of
+        // these things will get you some kind of Node installation:
+        // - A valid `package.json` even if it doesn't contain versions.
+        // - The presence of a `yarn.lock` file.
+        // - The presence of a `.nvmrc` file.
+        //
+        // TODO(zmitchell, 2025-08-01): the previous iteration of this setup
+        // code didn't really use `package-lock.json` to inform whether we
+        // were in a good state, so this doesn't either, but we may decide
+        // to do that in the future.
+        self.package_json_versions.is_some()
+            || self.yarn_lock_exists
+            || self.nvmrc_contents.as_ref().is_ok_and(|c| c.is_some())
+    }
+
+    /// Returns whether we should install both `yarn` and a matching `node`.
+    fn should_install_yarn_and_node(&self) -> bool {
+        // There may be some cases missing from this logic, but I *think* this
+        // reproduces the existing behavior.
+        self.yarn_lock_exists
+            && self
+                .package_json_versions
+                .as_ref()
+                .is_some_and(|versions| versions.yarn.is_some())
+            && self
+                .package_json_versions
+                .as_ref()
+                .is_some_and(|versions| versions.node.is_some())
+            && !self.package_lock_exists
+    }
+
+    /// Returns whether we should install just `yarn` without `node`.
+    fn should_install_yarn_only(&self) -> bool {
+        // If both `yarn.lock` and `package-lock.json` exist, we're in an
+        // ambiguous state since these files serve the same purpose.
+        // It could mean that the user just left one of them around after
+        // switching between package managers. In this ambiguous state
+        // we defer to the `package-lock.json`
+        self.yarn_lock_exists && self.package_json_versions.is_some() && !self.package_lock_exists
+    }
+
+    /// Returns whether we should only install `node`.
+    fn should_install_node_only(&self) -> bool {
+        // The bar here is pretty low
+        self.package_json_versions.is_some() || self.nvmrc_contents.is_ok()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum NodeVersionSource {
```

Comment:
This overlaps a little bit with `Node` itself, but I figured I had already rewritten enough of the world. This captures which source we tried to get a version from and then how successful we were getting a resolved package from that version information. This is primarily used in the messaging functionality.